internal/clientconn/connmetrics/conn_metrics.go

```diff
@@ -0,0 +1,183 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connmetrics
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// ConnMetrics represents conn metrics.
+type ConnMetrics struct {
+	Requests          *prometheus.CounterVec
+	Responses         *prometheus.CounterVec
+	AggregationStages *prometheus.CounterVec
+
+	cmds []string
+}
+
+type CommandMetrics interface{}
+
+type BasicCommandMetrics struct {
+	Failed int64
+	Total  int64
+}
+
+type UpdateCommandMetrics struct {
+	ArrayFilters int64
+	Failed       int64
+	Pipeline     int64
+	Total        int64
+}
+
+// newConnMetrics creates new conn metrics.
+//
+// The cmds is the list of all expected commands that could be measured.
+// After providing them, they will be set with the zero values.
+func newConnMetrics(cmds []string) *ConnMetrics {
+	return &ConnMetrics{
+		Requests: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "requests_total",
+				Help:      "Total number of requests.",
+			},
+			[]string{"opcode", "command"},
+		),
+		Responses: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "responses_total",
+				Help:      "Total number of responses.",
+			},
+			[]string{"opcode", "command", "result"},
+		),
+		AggregationStages: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "aggregation_stages_total",
+				Help:      "Total number of aggregation pipeline stages.",
+			},
+			[]string{"command", "stage"},
+		),
+		cmds: cmds,
+	}
+}
+
+// Describe implements prometheus.Collector.
+func (cm *ConnMetrics) Describe(ch chan<- *prometheus.Desc) {
+	cm.Requests.Describe(ch)
+	cm.Responses.Describe(ch)
+	cm.AggregationStages.Describe(ch)
+}
+
+// Collect implements prometheus.Collector.
+func (cm *ConnMetrics) Collect(ch chan<- prometheus.Metric) {
+	cm.Requests.Collect(ch)
+	cm.Responses.Collect(ch)
+	cm.AggregationStages.Collect(ch)
+}
+
+func (cm *ConnMetrics) GetResponses() map[string]CommandMetrics {
+	res := map[string]CommandMetrics{}
+
+	// initialize commands in the map to show zero values in the metrics output
+	for _, cmd := range cm.cmds {
+		// update related operators have more fields in the output
+		switch cmd {
+		case "update", "clusterUpdate", "findAndModify":
+			res[cmd] = UpdateCommandMetrics{}
+		default:
+			res[cmd] = BasicCommandMetrics{}
+		}
+	}
+
+	ch := make(chan prometheus.Metric)
```

Comment:
If it won't collide with any imports, `metrics` would be a better name for that channel.

Reply:
you right